### ISSUE TITLE
ParallelWrapper odd param updates issue (gradient sharing)

### ISSUE THREAD

**Issue by crockpotveggies** [deeplearning4j/deeplearning4j#4270]
#### Issue Description

When training a seq2seq autoencoder for denoising, training on a single device yields the following training stats:

![screen shot 2017-11-08 at 2 11 08 pm](https://user-images.githubusercontent.com/1445295/32578293-89da5f02-c491-11e7-948e-0f1315df367e.png)

In the training process above, the score converges at around 0.09.

However, when using ParallelWrapper a strange updates pattern emerges. There seem to be a couple issues with the PW approach:

**Sometimes update parameter ratios display frequent NaN patterns when using a threshold of 1e-3**

![screen shot 2017-11-08 at 2 07 10 pm](https://user-images.githubusercontent.com/1445295/32578389-e2de7160-c491-11e7-8c51-3fdce9318873.png)

**When using a threshold of 1e-4 the score never converges lower than 0.25**

I'm not quite sure why that's the case. Either the single device training is overfitting (not likely considering dropout is used...?) or there's an issue with how the parameter updates are handled.

#### Version Information

Please indicate relevant versions, including, if relevant:

* Deeplearning4j version 0.9.1
* CUDA 8.0
* Titan X Pascal

**Comment by crockpotveggies** [comment]
Attaching implementation:

```
        ParallelWrapper wrapper = new ParallelWrapper.Builder(net)
                .prefetchBuffer(24)
                .workers(Nd4j.getAffinityManager().getNumberOfDevices())
                .trainerFactory(new SymmetricTrainerContext())
                .trainingMode(ParallelWrapper.TrainingMode.CUSTOM)
                .gradientsAccumulator(new EncodedGradientsAccumulator(2, 1e-5))
                .build();
```

**Comment by raver119** [comment]
> https://deeplearning4j.org/distributed
> http://nikkostrom.com/publications/interspeech2015/strom_interspeech2015.pdf

As explained here, gradients sharing works via accumulation of gradients. Only gradients with module greater or equal to threshold are shared, others are accumulated locally, and re-evaluated on next iteration(s). So, for each individual update step there's only few values within gradients update message: +threshold, -threshold and 0.0. Plus, if that's a dense update message, there's also -threshold/2 value possible. So, 0.0s in gradients aren't that unexpected. That's the part of the idea: we share only gradients that are big enough.
